### ISSUE TITLE
Stochastic depth

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn_common#99]
We should have this implemented.
The name varies a bit in related literature and other implementations. Sometimes it is called stochastic layers, sometimes drop path, layer drop, even drop connect although that usually means sth different (dropout on weights).

This was introduced in this paper: [Deep Networks with Stochastic Depth](https://arxiv.org/abs/1603.09382)

This is commonly used for Transformers. For example:
- [Very Deep Self-Attention Networks for End-to-End Speech Recognition](https://arxiv.org/abs/1904.13377)
- [Reducing Transformer Depth on Demand with Structured Dropout](https://arxiv.org/abs/1909.11556)
- [Swin Transformer: Hierarchical Vision Transformer using Shifted Windows](https://arxiv.org/pdf/2103.14030)

External example implementations:

- [quanpn90 / NMTGMinor (Very Deep Self-Attention Networks for End-to-End Speech Recognition)](https://github.com/quanpn90/NMTGMinor/blob/803730ffb264caa99fe5abca8d21e5d91fd40269/onmt/modules/StochasticTransformer/Layers.py#L43). Jointly applied to attention and FF layer.
- [torchvision](https://github.com/pytorch/vision/blame/main/torchvision/ops/stochastic_depth.py) ([PR](https://github.com/pytorch/vision/pull/4301/files))
- [rwightman / pytorch-image-models: Swin Transformer](https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/swin_transformer.py), [`DropPath`](https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/layers/drop.py#L155)

This needs to depend (conditional logic, #24) on the train flag (#18).
